### ISSUE TITLE
Copying Uint8Array from JS to R seems extremely slow

### ISSUE THREAD

**Issue by jeroen** [r-wasm/webr#203]
Benchmarking my data export tool, I found that the bottleneck is not R, but copying binary data from JavaScript to the virtual FS. The other way around (from R to JS) is 1000x faster, so maybe I am doing something wrong.

For example, lets create a `Uint8Array` of 10MB:

```js
const webr = require("@r-wasm/webr");
const session = new webr.WebR();
const buffer = new ArrayBuffer(10*1024*1024);
const array = new Uint8Array(buffer);
await session.init();
```

Writing this to the R virtual filesystem takes 8 seconds:

```js
await session.FS.writeFile('test.bin', array);
```

Trying to copy the data into an R raw vector even takes 40 seconds:

```js
const d = await new session.RRaw(array);
```

Am I doing something wrong? By comparison, allocating a 10MB vector in webr, writing this to the virtual filesystem, and then reading it in JavaScript, only 21ms so this seems not be the issue

```js
// this is super fast
await session.evalRVoid("writeBin(raw(10*1024*1024), 'test.bin')");
var outbuf = await session.FS.readFile('test.bin');
var res = Buffer.from(outbuf, 'binary');
```

Here is a full script to run in node:

```js
const webr = require("@r-wasm/webr");
const session = new webr.WebR();

async function benchmark(){
	const buffer = new ArrayBuffer(10*1024*1024);
	const array = new Uint8Array(buffer);
	const t0 = new Date();
	await session.init();
	console.log(`Copying ${array.length} bytes to FS`)
	const t1 = new Date();
	await session.FS.writeFile('test.bin', array);
	const t2 = new Date();
	var outbuf = await session.FS.readFile('test.bin');
	var res = Buffer.from(outbuf, 'binary');
	console.log(`Copied ${res.length} bytes back to js`)
	const t3 = new Date();
	console.log("Init: ", t1-t0, "ms");
	console.log("writeFile: ", t2-t1, "ms");
	console.log("readFile: ", t3-t2, "ms");
	session.close();
}

benchmark()

// Init:  873 ms
// writeFile:  9871 ms
// readFile:  15 ms
```

**Comment by georgestagg** [comment]
Thanks for digging into this and reporting back!

I _think_ I know where the issue lies: the communication channel between the webR main thread and worker thread uses a serialisation step where messages are encoded into a byte array for transmission (currently we are unable to use the more standard `PostMessage` and transferable object mechanism here, for technical reasons). The serialisation step is not particularly optimised, and I guess that it is choking on messages containing large `Uint8Array`s.

The messaging back from the worker thread to the main thread _is_ able to use `PostMessage` and transfer objects, so does not suffer from this issue. AFAIK transferring an object in that way doesn't even involve a copy, instead detaching the buffer from one JS thread and attaching it to another, explaining the significantly higher performance. I do not think we'll be able to reach the same performance in the other direction until we can avoid the serialisation step altogether.

I'll take a look at introducing a more efficient serialisation scheme and report back. Certainly it would be better to avoid re-encoding `Uint8Array` if possible.

**Comment by georgestagg** [comment]
Looks like a better serialisation scheme helps a lot. On my machine,

Before
------
```
$ node benchmark.js                                
Copying 10485760 bytes to FS
Copied 10485760 bytes back to js
Init:  648 ms
writeFile:  3155 ms
readFile:  11 ms
```

After
------
```
$ node benchmark.js
Copying 10485760 bytes to FS
Copied 10485760 bytes back to js
Init:  417 ms
writeFile:  204 ms
readFile:  6 ms
```

Still not as efficient as the other direction, but still well worth changing. I'll formalise the change into a PR now.

**Comment by jeroen** [comment]
Awesome thanks! This will help a lot.

>  I do not think we'll be able to reach the same performance in the other direction until we can avoid the serialisation step altogether.

I'm still curious about the technical reason that PostMessage works for `readFile` but `writeFile` has to serialize. Would there be some workaround possible where we allocate the virtual file buffer in webr, and share it with JS such that we can populate data in JavaScript but still read it in webr? Or is sharing memory in such a way prevented by the wasm sandbox?

**Comment by georgestagg** [comment]
> Would there be some workaround possible where we allocate the virtual file buffer in webr, and share it with JS such that we can populate data in JavaScript but still read it in webr?

Possibly, using `SharedArrayBuffer`. But, we'd have to expand the messaging and encode further state so that the main thread can tell first the worker thread how big of a buffer to allocate, wait for the shared buffer from the worker, and finally signal when it has finished populating it. I think this would be possible but quite tricky to get right.

Even so, I'm not sure what the equivalent would look like for webR users in the browser who can't use `SharedArrayBuffer` and are stuck with the service worker fallback.

---

> I'm still curious about the technical reason that PostMessage works for readFile but writeFile has to serialize.

When the R wasm process starts, it waits for input from the user (or for other webR messages) inside an R routine compiled to wasm. While we block inside wasm waiting for the user input, we cannot yield to the JS event loop[1]. So, any messages sent from the main thread to the worker thread using `PostMessage` are never received because the `MessageEvent` never fires.

Messages sent the other way (from R to JS) work fine because the main thread's event loop is not blocked in the same way.

Eventually, the issue should go away if/when [stack switching](https://github.com/WebAssembly/stack-switching) is accepted and implemented in Wasm engines, at which point we could go back to using `PostMessage` and transferring the buffers efficiently.

[1] Asyncify would allow us to yield, but it does not work well for us due to the way the R interpreter works and the large asyncify overhead induced.

**Comment by georgestagg** [comment]
We've merged #204, so I'm going to close this issue for now.

If performance continues to be a problem when sending messages under the new msgpack encoding, drop a comment here and I'll reopen the thread.
